File: src/Leitor/LeitorDisciplinaEstudante.java
package Leitor;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;

import Modelo.Disciplina;
import Modelo.Estudante;

public class LeitorDisciplinaEstudante extends ILeitor implements Serializable{

    private static final long serialVersionUID = 1L;
    HashMap<String, ArrayList<Estudante>> mapDisciplina;
    HashMap<String,ArrayList<Disciplina>> mapEstudante;
    static LeitorDisciplinaEstudante leitor;

    private LeitorDisciplinaEstudante(){
        leitor=null;
        mapDisciplina = new HashMap<>();
        mapEstudante = new HashMap<>();
    }

    public static void setInstancia(LeitorDisciplinaEstudante novo){
        leitor = novo;
    }

    public static LeitorDisciplinaEstudante obterInstancia(){
        if (leitor == null){
            leitor = new LeitorDisciplinaEstudante();
            return leitor;
        }
        else return leitor;
    }

    public void adiciona(Disciplina disciplina, Estudante estudante){
        if(!mapEstudante.containsKey(estudante.obterMatricula())){
            mapEstudante.put(estudante.obterMatricula(),new ArrayList<>()); 
        }
        mapEstudante.get(estudante.obterMatricula()).add(disciplina);

        if(!mapDisciplina.containsKey(disciplina.obterCodigo() + disciplina.obterPeriodo().obterCodigo())){
            mapDisciplina.put(disciplina.obterCodigo() + disciplina.obterPeriodo().obterCodigo(),new ArrayList<>());
        }
        mapDisciplina.get(disciplina.obterCodigo() + disciplina.obterPeriodo().obterCodigo()).add(estudante);
    }

    public Estudante busca(Disciplina disciplina,String matricula){
        if (!mapDisciplina.containsKey(disciplina.obterCodigo() + disciplina.obterPeriodo().obterCodigo()))
            return null;
        for(Estudante e : mapDisciplina.get(disciplina.obterCodigo() + disciplina.obterPeriodo().obterCodigo())){
            if( e.obterMatricula()==matricula)
                return e;
        }
        return null;
    }

    public ArrayList<Estudante> busca(Disciplina disciplina){
        return mapDisciplina.get(disciplina.obterCodigo() + disciplina.obterPeriodo().obterCodigo());
    }
    public ArrayList<Disciplina> busca(Estudante estudante){
        return mapEstudante.get(estudante.obterMatricula());
    }
}

File: cpp_version/Leitor/LeitorDisciplinaEstudante.cpp
#include "LeitorDisciplinaEstudante.h"

LeitorDisciplinaEstudante* LeitorDisciplinaEstudante::leitor = nullptr;

LeitorDisciplinaEstudante::LeitorDisciplinaEstudante(){}

LeitorDisciplinaEstudante* LeitorDisciplinaEstudante::obterInstancia(){
    if (leitor == NULL){
        leitor = new LeitorDisciplinaEstudante();
        return leitor;
    }
    else return leitor;
}

void LeitorDisciplinaEstudante::adiciona(Disciplina* disciplina, Estudante* estudante){
    bool contains = false;
    for(map<string,vector<Disciplina*>>::iterator it = mapEstudante.begin(); it != mapEstudante.end(); ++it){
        if(it->first == estudante->obterMatricula()){
            contains = true;
        }
    }
    if(!contains){
        vector<Disciplina*> novo;
        mapEstudante[estudante->obterMatricula()] = novo;
    }
    mapEstudante[estudante->obterMatricula()].push_back(disciplina);

    contains = false;
    for(map<string,vector<Estudante*>>::iterator it = mapDisciplina.begin(); it != mapDisciplina.end(); ++it){
        if(it->first == disciplina->obterCodigo() + disciplina->obterPeriodo()->obterCodigo()){
            contains = true;
        }
    }
    if(!contains){
        vector<Estudante*> novo;
        mapDisciplina[disciplina->obterCodigo() + disciplina->obterPeriodo()->obterCodigo()] = novo;
    }
    mapDisciplina[disciplina->obterCodigo() + disciplina->obterPeriodo()->obterCodigo()].push_back(estudante);
}

Estudante* LeitorDisciplinaEstudante::busca(Disciplina* disciplina,string matricula){
    bool contains = false;
    for(map<string,vector<Estudante*>>::iterator it = mapDisciplina.begin(); it != mapDisciplina.end(); ++it){
        if(it->first == disciplina->obterCodigo() + disciplina->obterPeriodo()->obterCodigo()){
            contains = true;
        }
    }
    if(!contains){
        return nullptr;
    }
    for(auto e : mapDisciplina[disciplina->obterCodigo() + disciplina->obterPeriodo()->obterCodigo()]){
        if(e->obterMatricula() == matricula)
            return e;
    }
    return nullptr;
}

vector<Estudante*> LeitorDisciplinaEstudante::busca(Disciplina* disciplina){
    return mapDisciplina[disciplina->obterCodigo() + disciplina->obterPeriodo()->obterCodigo()];
}

vector<Disciplina*> LeitorDisciplinaEstudante::busca(Estudante* estudante){
    return mapEstudante[estudante->obterMatricula()];
}